### PR TITLE
ENG-798: Benchmark state size

### PR DESCRIPTION
**Pull request by aakoshh**
Collect baseline statistics for state growth.

The PR adds a `fendermint/testing/benchmarks/state-size` directory with a README about how to collect growth statistics over a standalone node doing nothing but empty blocks. The statistics are saved under the `measurements` directory.

TODO: 
- [x] Collect measurements 
- [x] Visualize data
- [x] Print simple average

| App history | On-chain metadata  | Compaction | Collected |
| --- | :---: | :---: | :---: |
| Unlimited  | On | None | :heavy_check_mark: |
| Limited | On  | None | :heavy_check_mark:  |
| Limited | Off | None | :heavy_check_mark:  |
| Limited | Off | Level | :heavy_check_mark:  |

### Preliminary results

Looking at the collected statistics, it looks like if on-chain metadata is enabled (that is, the block hash of the current epoch is stored in the ledger), then each block height adds ~5KB of data to the storage, which, with 1s block time, results in a daily growth of ~430MB. 

With on-chain metadata disabled, the growth is only ~0.6KB per block height.

I am not familiar with the implementation of the AMT we use to store the metadata, but with the HAMT it was storing by default 32 key-value entries in each node before it started branching (and then 32 CIDs). These nodes had to be copied if any of the slots changed. 

The AMT seems to use a different bit width, only 3, which I assume means 8 slots. We store a sliding window of 256 epochs, which I assume applies a tree structure with at height of at least 3, so writing a new value could result in writing a node of `8*32` bytes on the leaf and then `2*8*64` bytes for CIDs on higher levels (IIRC a CID is 60 something bytes), which is 1280 bytes. This ignores the fact that the AMT is a sparse array, so it has to store the epochs as well. In addition, we will have to modify the actor storages, which are in HAMTs. I guess it could add up to >4KB.

### Followup

Dig deeper into:
* The AMT implemenation, see what the net amount of bytes we write is, to check if there is any overhead from the blockstore itself
* Find out why, with a limited app-state history size and metadata disabled, we still see linear growth in database size. Why isn't RocksDB reusing the space? Will it ever reuse it? Do we have to compact it somehow?
* Garbage collection should free up the unused AMT roots, but still we also need to understand whether in-place deletion in RocksDB will result in reclaimed space.

### Compaction

https://github.com/facebook/rocksdb/wiki/Compaction explains the different compaction styles. Currently fendermint uses [no compaction](https://github.com/consensus-shipyard/ipc/blob/14fc07120e6b1b0e8884653d9eb1b632d95d1acf/fendermint/rocksdb/src/rocks/config.rs#L40) which [disables auto compaction](https://github.com/consensus-shipyard/ipc/blob/14fc07120e6b1b0e8884653d9eb1b632d95d1acf/fendermint/rocksdb/src/rocks/config.rs#L67). We can try adding this to the configuration and enabling it.

I set the default to `level` but not sure when it happens. Need to dig into https://github.com/facebook/rocksdb/wiki/Leveled-Compaction for options. Maybe manually running (which I don't know how to do) is worth exploring for the purpose of these benchmarks.

### CODE REVIEW

**Comment by aakoshh** [comment]
Took the ability to disable pushing meta-data from https://github.com/consensus-shipyard/ipc/pull/680

**Review by raulk — APPROVED** [review]
Thanks for adding this benchmark suite. Kudos for another successful use of the Materializer. All of this will come in extremely handy for optimization.

In terms of takeaways from the discoveries, how would we feel about:

1. Prototying the removal of the chain metadata actor and instead keeping the lookback in app state, or
2. Using the AMT as a ring buffer instead of how we use it today (linear FIFO queue)

@aakoshh are you able to write another test scenario that sends a simple transfer transaction with `push_chain_meta`? That would isolate the cost of updating the state tree to help us choose between (1) or (2) as a potential next step.

**Review by raulk — COMMENTED** [review]
On compation: based on my understanding of LSM trees, enabling compation will have no net effect unless our workloads are mutating or deleting KV entries. IUC, we aren't doing that just now and won't until we implement state-tree level GC.

**Comment by aakoshh** [comment]
> Prototying the removal of the chain metadata actor and instead keeping the lookback in app state

Whenever we want to move things into the app state we have to consider snapshotting. Currently the only [state related field](https://github.com/consensus-shipyard/ipc/blob/14fc07120e6b1b0e8884653d9eb1b632d95d1acf/fendermint/vm/snapshot/src/manifest.rs#L31) in the `SnapshotManifest` is the `FvmStateParams` itself; similarly the [AppState](https://github.com/consensus-shipyard/ipc/blob/14fc07120e6b1b0e8884653d9eb1b632d95d1acf/fendermint/app/src/app.rs#L73-L78) and its [history](https://github.com/consensus-shipyard/ipc/blob/14fc07120e6b1b0e8884653d9eb1b632d95d1acf/fendermint/app/src/app.rs#L276-L277) is just `FvmStateParams` records, which means it has to be added to that, or a lot of refactoring has to happen. 

Given that the reason we do this is to serve the `BLOCKHASH` function of the EVM actor, I think it makes sense that this state is either in the ledger, or the `FvmStateParams`, catered for the `FvmMessageInterpreter`, not the `App` itself.

I thought it made sense that `FvmStateParams` only contains data that the FVM executor requires for its instantiation (timestamp, block height), and everything else goes in the ledger. For example the `Init` actor is a good example: it could be outside, in application state, because we know we'll have actors, it's non-optional, but instead it is implemented as a built-in actor. 

The `FvmStateParams` record is copied as-is from slot to slot in the history, so ideally its size should be kept small. Storing `256*32` bytes and copying it as a single record from slot to slot would mean 8KB of data, which is double the growth of what we are looking at now, and we also made this a configurable parameter, so it can be more.


So, we can have a story of storing more chain metadata in the `App`, e.g. the validator set, the block hashes, stuff we sometimes need during block execution, but we know CometBFT is not always available to get it, but we have to be aware that to make these part of the snapshots requires extra work, which for ledger stored data comes for free. I'm not sure storage size saving is a good motivation for this; if we had garbage collection take care of this, it wouldn't matter where the data is from this perspective, save for some constant overhead.

**Comment by aakoshh** [comment]
> Using the AMT as a ring buffer instead of how we use it today (linear FIFO queue)

I am not that familiar with how the AMT works internally, but I imagine due to the immutability of the nodes that build up the tree, we'll end up with copying anyway. 

For example if we had a tree that looks like this:

```
   Root [Cid(Leaf1), Cid(Leaf2)]
               /           \
        Leaf1 [0-7]       Leaf2 [8-15]
```

When the time comes to insert the next height, we have two options:
* Create a new `Leaf3` and insert a copy of `Root` with a new CID appended to it, or
* Modify `Leaf1` to replace index 0 with the new data, which involves making a modified copy of `Leaf1` and `Root`. 

Overall it's not obvious which one would have a better amortised cost.

**Comment by aakoshh** [comment]
> are you able to write another test scenario that sends a simple transfer transaction with `push_chain_meta`? That would isolate the cost of updating the state tree

So you want to know, with `push_chain_meta` _enabled_, what additional cost it would incur to update the actor balance of two accounts?

File: fendermint/app/config/default.toml [hunk 1]
```diff
@@ -95,6 +97,21 @@ gas_search_step = 1.25
...
+#
+# This is here for testing purposes only, it should be `true` by default to allow
+# the `evm` actor to execute the `BLOCKHASH` function. If you want to disable
+# empty blocks created by CometBFT, for a less frenetic testing experience,
+# then you need to disable this as well, with the side effect that `BLOCKHASH`
+# will always return an empty value.
+#
+# See https://docs.cometbft.com/v0.37/core/configuration#empty-blocks-vs-no-empty-blocks
+push_chain_meta = true
+
```
[the remark this answers is not in the document]
Comment (aakoshh) [reply fendermint/app/config/default.toml:114]: I can move it around, but there's currently no way to tell whether you start fendermint "for production". There is the `NETWORK` setting but I never assign it, all e2e tests use the default `main` with f410 addresses for example. Did you have anything in mind which would deny the ability to set test settings?

[the remark this answers is not in the document]
Comment (aakoshh) [reply fendermint/app/config/default.toml:114]: I can tie it to the the `NETWORK` being test, it should be easy to set it in this particular scenario, and nothing else needs to change.

**Comment by aakoshh** [comment]
> enabling compation will have no net effect unless our workloads are mutating or deleting KV entries.

The only deletion we do is the clearing out of old state history records [here](https://github.com/consensus-shipyard/ipc/blob/14fc07120e6b1b0e8884653d9eb1b632d95d1acf/fendermint/app/src/app.rs#L283), which should allow the state roots to be garbage collected, when it's implemented. You are right that in itself this is unlikely to trigger compaction. I hope there is no adverse affect in enabling an auto-compaction, though.

**Comment by raulk** [comment]
> > are you able to write another test scenario that sends a simple transfer transaction with `push_chain_meta`? That would isolate the cost of updating the state tree
> 
> So you want to know, with `push_chain_meta` _enabled_, what additional cost it would incur to update the actor balance of two accounts?

Sorry, it looks like I didn't finish my sentence. Meant to say: "with `push_chain_meta=false`" Right now we're comparing:

1. The cost of updating the state tree with the pushed chain metadata.
2. The cost of NOT updating the state tree at all.

To get to the net cost of _just_ pushing the chain metadata, we need to either (a) instrument those writes, or (b) do a cheap state-mutating operation like a transfer and subtract that cost from the total. The latter seems less work.

File: fendermint/app/config/default.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (raulk) [reply fendermint/app/config/default.toml:114]: How about using the presence of the `RUST_TEST_THREADS` env variable to determine if we're running or not in a cargo test? From what I gather, it's always non-zero within a test.

**Comment by aakoshh** [comment]
I tested the transfer by burning 1 atto from the validator account with the following code added to `exec.rs` in `begin_block`: 

```rust
        if let Some(validator_id) = state.validator_id() {
            // TODO: Make it more efficient to turn an ID into a PublicKey.
            let (_, power_table) = self
                .gateway
                .current_power_table(&mut state)
                .context("failed to get the power table")?;

            let validator = power_table.iter().find(|v| {
                if let Ok(pk) = tendermint::PublicKey::try_from(v.public_key.clone()) {
                    let id = tendermint::account::Id::from(pk);
                    id == validator_id
                } else {
                    false
                }
            });

            if let Some(v) = validator {
                // Burn some funds from the validator.
                let msg = FvmMessage {
                    from: Address::from(EthAddress::from(v.public_key.0)),
                    to: burntfunds::BURNT_FUNDS_ACTOR_ADDR,
                    sequence: height as u64,
                    gas_limit,
                    method_num: METHOD_SEND,
                    params: Default::default(),
                    value: TokenAmount::from_atto(1),
                    version: Default::default(),
                    gas_fee_cap: Default::default(),
                    gas_premium: Default::default(),
                };

                let (apply_ret, _) = state.execute_implicit(msg)?;

                if let Some(err) = apply_ret.failure_info {
                    anyhow::bail!("failed to burn funds from block creator: {}", err);
                }
            }
        }
```
Originally wanted to make it into a reward, but forgot that not even the system actor can mint funds like this.

Anyway, @raulk you were right, I didn't run the experiment more than a few minutes but the average growth is 4KB, so changing the actor state in itself accounts for most of the 5KB we saw in the beginning with only the metadata.

**Pull request merged**